src/transform/src/fusion/reduce.rs

```diff
@@ -0,0 +1,105 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Fuses reduce operators with parent operators if possible.
+
+use crate::TransformArgs;
+use expr::{AggregateExpr, AggregateFunc, MirRelationExpr, MirScalarExpr};
+
+/// Fuses reduce operators with parent operators if possible.
+#[derive(Debug)]
+pub struct Reduce;
+
+impl crate::Transform for Reduce {
+    fn transform(
+        &self,
+        relation: &mut MirRelationExpr,
+        _: TransformArgs,
+    ) -> Result<(), crate::TransformError> {
+        relation.visit_mut_pre(&mut |e| {
+            self.action(e);
+        });
+        Ok(())
+    }
+}
+
+impl Reduce {
+    /// Fuses reduce operators with parent operators if possible.
+    pub fn action(&self, relation: &mut MirRelationExpr) {
+        if let MirRelationExpr::Reduce {
+            input,
+            group_key,
+            aggregates,
+            monotonic: _,
+            expected_group_size: _,
+        } = relation
+        {
+            if let MirRelationExpr::Reduce {
+                input: inner_input,
+                group_key: inner_group_key,
+                aggregates: inner_aggregates,
+                monotonic: _,
+                expected_group_size: _,
+            } = &mut **input
+            {
+                // Do nothing if outer keys are not a subset of inner keys
+                if !group_key.iter().all(|k| inner_group_key.contains(k)) {
+                    return;
+                }
+
+                if aggregates.is_empty() && inner_aggregates.is_empty() {
+                    // Drop redundant inner reduce
+                    *input = Box::new(inner_input.take_dangerous());
+                } else if aggregates.len() == 1 && inner_aggregates.len() == 1 {
+                    // TODO: This can be more general than just len() == 1
```

Comment:
I didn't think too hard about this, but I think we could extend this. I first wanted to hear back whether this looks generally OK before handling more cases. I'm happy to do a follow-up.

Reply:
I'd limit the scope of the first patch to the cases where the outer reduction doesn't depend at all on the result of any aggregation performed by the inner reduction.

In a second patch I'd relax that condition to check that all aggregations in the outer reduction can be rewritten in a way that doesn't depend on the result of a previous aggregation.

Also, it doesn't matter how many aggregations the inner reduction has as long as all the outer ones can be expressed in such a way that they depend only on input columns.

Reply:
> I'd limit the scope of the first patch to the cases where the outer reduction doesn't depend at all on the result of any aggregation performed by the inner reduction.

I think that's wise and I was also mildly annoyed with myself for mixing these two things in the first PR. I've limited this PR to the case where the outer reduction doesn't depend at all on the result of any aggregation performed by the inner reduction.